File: server/app/schemas/confluence.py
from datetime import datetime
from typing import Optional, List, Dict, Any, Literal
from pydantic import BaseModel, Field, HttpUrl
from uuid import UUID

from app.models import ConfluenceType, ImportType, SyncStatus


# Authentication Schemas
class ConfluenceAuthRequest(BaseModel):
    confluence_type: ConfluenceType
    base_url: str
    email: Optional[str] = None
    api_token: Optional[str] = None
    oauth_token: Optional[str] = None
    oauth_refresh_token: Optional[str] = None
    token_expires_at: Optional[datetime] = None

class ConfluenceAuthResponse(BaseModel):
    id: UUID
    confluence_type: ConfluenceType
    base_url: str
    email: Optional[str]
    created_at: datetime
    updated_at: datetime
    is_valid: bool

class ConfluenceCredentialList(BaseModel):
    credentials: List[ConfluenceAuthResponse]


# Space and Page Schemas
class ConfluenceSpace(BaseModel):
    key: str
    name: str
    description: Optional[str] = None
    homepage_id: Optional[str] = None

class ConfluencePage(BaseModel):
    id: str
    title: str
    space_key: str
    parent_id: Optional[str] = None
    version: int
    created_date: Optional[datetime] = None
    modified_date: Optional[datetime] = None
    has_children: bool = False
    has_attachments: bool = False


# Import Schemas
class ConfluenceImportRequest(BaseModel):
    credential_id: UUID
    folder_id: UUID
    import_type: ImportType
    space_key: Optional[str] = None
    page_id: Optional[str] = None
    include_attachments: bool = True
    include_comments: bool = False
    recursive: bool = True

class ConfluenceImportResponse(BaseModel):
    id: UUID
    import_type: ImportType
    sync_status: SyncStatus
    total_pages: int = 0
    processed_pages: int = 0
    error_message: Optional[str] = None
    created_at: datetime
    last_sync_at: Optional[datetime] = None

class ConfluenceImportStatus(BaseModel):
    id: UUID
    status: SyncStatus
    total_pages: int
    processed_pages: int
    error_message: Optional[str] = None
    created_at: datetime
    last_sync_at: Optional[datetime] = None
    progress_percentage: float = Field(default=0.0)
    
    @property
    def progress_percentage(self) -> float:
        if self.total_pages == 0:
            return 0.0
        return (self.processed_pages / self.total_pages) * 100


# Search and Discovery Schemas
class ConfluenceSearchRequest(BaseModel):
    credential_id: UUID
    query: str
    search_type: Literal["cql", "text"] = "text"
    limit: int = Field(default=50, ge=1, le=100)

class ConfluenceSearchResult(BaseModel):
    id: str
    title: str
    space_key: str
    space_name: str
    excerpt: Optional[str] = None
    url: Optional[str] = None
    last_modified: Optional[datetime] = None


# Sync Schemas
class ConfluenceSyncRequest(BaseModel):
    import_id: UUID
    force_full_sync: bool = False

class ConfluenceSyncResponse(BaseModel):
    import_id: UUID
    status: str
    pages_updated: int
    pages_added: int
    pages_deleted: int
    sync_started_at: datetime
    sync_completed_at: Optional[datetime] = None


# Batch Import Schema
class ConfluenceBatchImportRequest(BaseModel):
    credential_id: UUID
    folder_id: UUID
    imports: List[Dict[str, Any]]
    
class ConfluenceBatchImportItem(BaseModel):
    import_type: ImportType
    space_key: Optional[str] = None
    page_id: Optional[str] = None

class ConfluenceBatchImportResponse(BaseModel):
    job_ids: List[UUID]
    total_jobs: int
    status: str

File: client/src/types/confluence.ts
export type ConfluenceType = "cloud" | "server" | "data_center";
export type ImportType = "page" | "space" | "page_tree";
export type SyncStatus = "pending" | "in_progress" | "completed" | "failed" | "partial";

export interface ConfluenceCredential {
  id: string;
  confluence_type: ConfluenceType;
  base_url: string;
  email: string | null;
  created_at: string;
  updated_at: string;
  is_valid: boolean;
}

export interface ConfluenceAuthRequest {
  confluence_type: ConfluenceType;
  base_url: string;
  email?: string;
  api_token?: string;
  oauth_token?: string;
  oauth_refresh_token?: string;
  token_expires_at?: string;
}

export interface ConfluenceSpace {
  key: string;
  name: string;
  description?: string;
  homepage_id?: string;
}

export interface ConfluencePage {
  id: string;
  title: string;
  space_key: string;
  parent_id?: string;
  version: number;
  created_date?: string;
  modified_date?: string;
  has_children: boolean;
  has_attachments: boolean;
}

export interface ConfluenceSearchRequest {
  credential_id: string;
  query: string;
  search_type?: "cql" | "text";
  limit?: number;
}

export interface ConfluenceSearchResult {
  id: string;
  title: string;
  space_key: string;
  space_name: string;
  excerpt?: string;
  url?: string;
  last_modified?: string;
}

export interface ConfluenceImportRequest {
  credential_id: string;
  folder_id: string;
  import_type: ImportType;
  space_key?: string;
  page_id?: string;
  include_attachments?: boolean;
  include_comments?: boolean;
  recursive?: boolean;
}

export interface ConfluenceImport {
  id: string;
  import_type: ImportType;
  sync_status: SyncStatus;
  total_pages: number;
  processed_pages: number;
  error_message?: string;
  created_at: string;
  last_sync_at?: string;
  space_key?: string;
  page_id?: string;
  folder_id: string;
}

export interface ConfluenceImportStatus {
  id: string;
  status: SyncStatus;
  total_pages: number;
  processed_pages: number;
  error_message?: string;
  created_at: string;
  last_sync_at?: string;
  progress_percentage: number;
}

export interface ConfluenceBatchImportRequest {
  credential_id: string;
  folder_id: string;
  imports: Array<{
    import_type: ImportType;
    space_key?: string;
    page_id?: string;
  }>;
}

export interface ConfluenceBatchImportResponse {
  job_ids: string[];
  total_jobs: number;
  status: string;
}

export interface ConfluenceSyncRequest {
  import_id: string;
  force_full_sync?: boolean;
}

export interface ConfluenceSyncResponse {
  import_id: string;
  status: string;
  pages_updated: number;
  pages_added: number;
  pages_deleted: number;
  sync_started_at: string;
  sync_completed_at?: string;
}